dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/executor/enrollment/SetMandatoryFieldExecutor.java

```diff
@@ -57,18 +63,41 @@ public Optional<ProgramRuleIssue> executeRuleAction(TrackerBundle bundle, Enroll
     TrackerIdSchemeParams idSchemes = bundle.getPreheat().getIdSchemes();
     TrackedEntityAttribute ruleAttribute =
         bundle.getPreheat().getTrackedEntityAttribute(attributeUid.getValue());
-    Optional<Attribute> any =
+
+    Set<MetadataIdentifier> enrollmentAttributes =
         enrollment.getAttributes().stream()
-            .filter(attribute -> attribute.getAttribute().isEqualTo(ruleAttribute))
+            .map(Attribute::getAttribute)
+            .collect(Collectors.toSet());
+
+    Set<MetadataIdentifier> teAttributes =
+        ValidationUtils.buildTeAttributes(bundle, enrollment.getTrackedEntity());
+
+    Optional<MetadataIdentifier> any =
+        Stream.concat(enrollmentAttributes.stream(), teAttributes.stream())
+            .filter(a -> a.isEqualTo(ruleAttribute))
             .findAny();
 
-    if (any.isEmpty() || StringUtils.isEmpty(any.get().getValue())) {
+    if (any.isEmpty() && bundle.getStrategy(enrollment) == TrackerImportStrategy.CREATE) {
       return Optional.of(
           error(
               ruleUid,
               ValidationCode.E1306,
               idSchemes.toMetadataIdentifier(ruleAttribute).getIdentifierOrAttributeValue()));
     }
+
+    Optional<Attribute> enrollmentAttribute =
+        enrollment.getAttributes().stream()
```

Comment:
On CREATE you
* check if the mandatory attribute **UID** is in TE or EN attributes. You do not check that its value is not null
* in case the mandatory attribute UID is present in the TE this code will execute. Why is it then only checking the value of the enrollment attribute? 🤔

Reply:
I blame the attribute model again :grimacing: 
Program rules are applied to an enrollment, here we cannot really return an error on a tracked entity.
And program rules do not apply to tracked entities, so I guess we cannot really return an error for a tracked entity that was calculated by a program rule.

This is controversial and to be honest I am not 100% sure myself if this is all we can do here

Reply:
but why do we take the TE into account when checking if the attribute UID is present? is that because we allow users to put a program attribute into the TE or EN?